lucetc/src/decls.rs

```diff
@@ -267,25 +267,42 @@ impl<'a> ModuleDecls<'a> {
         for ix in 0..info.globals.len() {
             let ix = GlobalIndex::new(ix);
             let g_decl = info.globals.get(ix).unwrap();
-            let g_import = info.imported_globals.get(ix);
-            let g_variant = if let Some((module, field)) = g_import {
-                GlobalVariant::Import { module, field }
-            } else {
-                let init_val = match g_decl.entity.initializer {
-                    // Need to fix global spec in ModuleData and the runtime to support more:
-                    GlobalInit::I32Const(i) => i as i64,
-                    GlobalInit::I64Const(i) => i,
-                    _ => Err(format_err!(
-                        "non-integer global initializer: {:?}",
-                        g_decl.entity
-                    ))
-                    .context(LucetcErrorKind::Unsupported)?,
-                };
-                GlobalVariant::Def {
-                    def: GlobalDef::new(init_val),
+
+            let global = match g_decl.entity.initializer {
+                GlobalInit::I32Const(i) => Ok(GlobalVariant::Def(GlobalDef::I32(i))),
+                GlobalInit::I64Const(i) => Ok(GlobalVariant::Def(GlobalDef::I64(i))),
+                GlobalInit::F32Const(f) => {
+                    Ok(GlobalVariant::Def(GlobalDef::F32(f32::from_bits(f))))
                 }
-            };
-            globals.push(GlobalSpec::new(g_variant, None));
+                GlobalInit::F64Const(f) => {
+                    Ok(GlobalVariant::Def(GlobalDef::F64(f64::from_bits(f))))
+                }
+                GlobalInit::GetGlobal(ref_ix) => {
+                    let ref_decl = info.globals.get(ref_ix).unwrap();
+                    if let GlobalInit::Import = ref_decl.entity.initializer {
+                        if let Some((module, field)) = info.imported_globals.get(ref_ix) {
+                            Ok(GlobalVariant::Import { module, field })
+                        } else {
+                            Err(format_err!("inconsistent state: global {} is declared as an import but has no entry in imported_globals", ref_ix.as_u32()))
+                            .context(LucetcErrorKind::TranslatingModule)
+                        }
+                    } else {
+                        // This WASM restriction may be loosened in the future:
+                        Err(format_err!("invalid global declarations: global {} is initialized by referencing another global value, but the referenced global is not an import", ix.as_u32()))
+                        .context(LucetcErrorKind::TranslatingModule)
+                    }
+                }
+                GlobalInit::Import => {
+                    if let Some((module, field)) = info.imported_globals.get(ix) {
+                        Ok(GlobalVariant::Import { module, field })
```

Comment:
This turns import globals into a runtime error for the moment. It probably makes sense to run `module` and `field` through the same name mapping as functions (eg `bindings.json`) but aside from spec tests it seems like the most common use of globals would be to import from other modules, which is also something lucet-runtime does not yet support.

Reply:
Yes, we need to figure out the design changes to lucet-runtime for import globals, memories, tables etc - right now import functions is special cased to use the linker, which is also not always possible or useful. So I think we should continue to defer that to a runtime error. This looks like the correct implementation for the compiler.

Reply:
I totally imagine an argument for embedding-provided globals, so there might be a path where this grows in the same way `ImportFunction` has a new field in #203. But maybe by then we'll support loading multiple wasm modules, and embedder context is just a wasm-module-looking thing. Ah, the problems of tomorrow!